compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertiesToFieldsLowering.kt

```diff
@@ -90,7 +115,7 @@ class PropertiesToFieldsLowering(val context: CommonBackendContext) : IrElementT
     }
 
     private fun substituteGetter(irProperty: IrProperty, expression: IrCall): IrExpression {
-        val backingField = irProperty.backingField!!
+        val backingField = irProperty.backingField!!.resolveFakeOverride()!!.getRealBackingField()
```

Comment:
Generally, we do not resolve fake override at compilation time, and I prefer if we didn't do it here as well. The reason is that if a member is moved to a subclass and the clients are not recompiled, in some circumstances the call site can end up calling the incorrect member.

It's not directly applicable to the case of `@JvmField` properties though, because they're always final and moving them is a binary incompatible change anyway. But I think it's still better to keep the old backend's behavior here, at least for some possible obscure migration cases.

Reply:
Hm, I didn't know that JVM allows referencing a field through a subclass (or maybe I had simply forgotten...) Resolving the fake override only for the check should be sufficient then (and safe, since objects are final).

Reply:
Speaking of fake overrides: the FIR team seems reluctant to add them, and we're also having trouble figuring out exactly what their purpose is. So far it seems like they're only really necessary when the type of a member changes, e.g. when inheriting from an instantiation of a generic type, as then some bridges might be required. For the case you've mentioned it seems sufficient to always reference members through the dispatch receiver's type (when non-static and not a `super` access) and repurpose `superQualifierSymbol` to mean "statically resolved target" (so it would point to some subtype of the field/method's owner if it's static). Is there something we're missing?

Reply:
You're not missing anything, and I also think it'd be better to get rid of fake overrides in the IR. It's not immediately obvious how to do this though, since they're serialized to the klib format (see `KotlinIr.proto`) and the deserializer relies on their existence to properly link call sites. We're discussing this issue right now and I hope we can come up with a plan, but it's not an easy refactoring.